acpower/acpower_benchmark.md

```diff
@@ -0,0 +1,255 @@
+          bus_voltage[k]^2*Gself(k),
+           bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def)
+
+        Zygote.gradient((bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def) -> bus[k].q_gen - bus[k].q_load -
+          if !isempty(in_lines[k])
+            sum( bus_voltage[k] * bus_voltage[branch[i].from] *
+                (Gin(i) * sin(bus_angle[k] - bus_angle[branch[i].from]) -
+                 Bin(i) * cos(bus_angle[k] - bus_angle[branch[i].from])) for i in in_lines[k] ) else 0 end -
+          if !isempty(out_lines[k])
+            sum( bus_voltage[k] * bus_voltage[branch[i].to] *
+                (Gout(i) * sin(bus_angle[k] - bus_angle[branch[i].to]) -
+                 Bout(i) * cos(bus_angle[k] - bus_angle[branch[i].to])) for i in out_lines[k] ) else 0 end
+             + bus_voltage[k]^2*Bself(k), bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def)
+    end
+    
+    for k in 1:nbus
+        Zygote.gradient((bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def) -> bus[k].q_load +
+          if !isempty(in_lines[k])
+            sum( bus_voltage[k] * bus_voltage[branch[i].from] *
+                (Gin(i) * sin(bus_angle[k] - bus_angle[branch[i].from]) -
+                 Bin(i) * cos(bus_angle[k] - bus_angle[branch[i].from])) for i in in_lines[k] ) else 0 end +
+          if !isempty(out_lines[k])
+            sum( bus_voltage[k] * bus_voltage[branch[i].to] *
+                (Gout(i) * sin(bus_angle[k] - bus_angle[branch[i].to]) -
+                 Bout(i) * cos(bus_angle[k] - bus_angle[branch[i].to])) for i in out_lines[k] ) else 0 end +
+            - bus_voltage[k]^2*Bself(k) - bus[k].q_max, bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def)
+
+        Zygote.gradient((bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def) -> bus[k].p_load +
+          if !isempty(in_lines[k])
+            sum( bus_voltage[k] * bus_voltage[branch[i].from] *
+                (Gin(i) * cos(bus_angle[k] - bus_angle[branch[i].from]) +
+                 Bin(i) * sin(bus_angle[k] - bus_angle[branch[i].from])) for i in in_lines[k] ) else 0 end +
+          if !isempty(out_lines[k])
+            sum( bus_voltage[k] * bus_voltage[branch[i].to] *
+                (Gout(i) * cos(bus_angle[k] - bus_angle[branch[i].to]) +
+                 Bout(i) * sin(bus_angle[k] - bus_angle[branch[i].to])) for i in out_lines[k] ) else 0 end  +
+            + bus_voltage[k]^2*Gself(k) - p_gen_upper*bus[k].p_gen, bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def)
+    end
+end
+
+@benchmark eval_constraint_jacobian_zygote(bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def)
+```
+
+```
+julia> @benchmark eval_constraint_jacobian_zygote(bus_voltage, bus_angle, bus_b_shunt, branch_tap, branch_def)
+BenchmarkTools.Trial:
+  memory estimate:  1.46 GiB
+  allocs estimate:  6210520
+  --------------
+  minimum time:     1.565 s (15.55% GC)
```

Comment:
Very interesting numbers. A bit disappointing that Zygote is 500x slower than JuMP for computing the jacobian...

Reply:
I was surprised by this myself, especially since the other benchmark showed some promise. Perhaps I made a mistake in a rush to meet the application deadline. I'll have a look at this to see if it might have been a user error.